[API] Expose `observeEncryptedLogsUploadResult` Flow

ParaskP7 opened the pull request:
Closes: [AINFRA-981](https://linear.app/a8c/issue/AINFRA-981/update-encrypted-logging-library-to-expose)
Internal Discussion: p1752847242612859-slack-C04PWEZSYFL
Future Improvement: #25

---

### Description

This PR brings back the `observeEncryptedLogsUploadResult()` function to SDK, making it publicly available again, effectively reverting the change done in this 3c19605c132a5c19b50fc58ca1bb54639a534aa1 commit.

---

### Testing Steps

Using the [DOAndroid](https://github.com/bloom/DayOne-Android) or/and [PCAndroid](https://github.com/Automattic/pocket-casts-android) project, update `automattic-encryptedlogging` to `26-bc0777f7bed8a7cdf9b1e4213e597a9adf7b834b` (version which is based on this PR's [latest](https://github.com/Automattic/EncryptedLogging/pull/26/commits/bc0777f7bed8a7cdf9b1e4213e597a9adf7b834b) commit) and make sure that encrypted logging continues to work as expected:

1. Using `App Inspection` from AS find the `encrypted-logging.db` and its sole `EncryptedLogModel` table. Notice that the table is empty, there are no log to be sent.
2. Make the app crash. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could navigate to `Settings` -> `Developer` and then click `Crash The App`.
3. You might now notice a new db entry within `App Inspection` from AS on that `DETACHED` app (or not, it depends). Now, switch to the `ATTACHED` app and notice the db empty. This is because this db entry/log has been already sent to Sentry already, on app start and after the app crashed. To really notice the entry, trying crashing the app again, this time put the device on offline mode first. Doing that you'll be able to see the entry on the db, even when switching to the newly created `ATTACHED` app.
4. Open [Sentry](https://a8c.sentry.io) and navigate to `Issues`. Find the project you're testing for (for example `day-one-android` or/and `pocket-casts-android`) and check the latest `debug` exceptions. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could search for `Developer caused a crash!` and check the latest crash. Now navigate at the bottom and check the `uuid`. Verify it is the same `UUID` that you saw on that db entry before.
5. Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that `UUID` into `Log UUID`, click `View` and make sure you can read the newly uploaded and now decrypted log.

---

Additionally, make sure that the added back `observeEncryptedLogsUploadResult()` works as expected. For example, for [DOAndroid](https://github.com/bloom/DayOne-Android), you could make sure that this [encryptedLogging.observeEncryptedLogsUploadResult().collect { ... }](https://github.com/bloom/DayOne-Android/blob/trunk/app/src/main/java/com/dayoneapp/dayone/utils/EncryptedLoggingInitializer.kt#L22-L39) logic still works as expected.

ParaskP7 committed:
Sdk: Bring back observe encrypted logs upload result function to sdk

Internal Discussion: p1752847242612859-slack-C04PWEZSYFL

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/OnChanged.kt:
```diff
@@ -1,8 +1,8 @@
-internal open class OnChanged<T : OnChangedError> {
-    var error: T? = null
-    fun isError(): Boolean {
+public open class OnChanged<T : OnChangedError> {
+    public var error: T? = null
+    public fun isError(): Boolean {
         return error != null
     }
```
As this is now part of the public contract, can we remove this to reduce classes exposed to clients? There's no value in this abstraction, it's just a leftover from FluxC.

ParaskP7 replied:
True @wzieba and thanks for the suggestion, I'll see what I can do! 👍

ParaskP7 replied:
Done: 9f037323f39aa0fc80310fea975c9d683f8a79a9 @wzieba

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/EncryptedLogging.kt:
```diff
@@ -25,6 +27,11 @@ public interface EncryptedLogging {
+    /**
+     * A method for the client to use to observe the upload result of the encrypted logs.
+     */
+    public fun observeEncryptedLogsUploadResult(): StateFlow<OnEncryptedLogUploaded?>
```
I don't think we have to expose `StateFlow` - it's an implementation detail.

```suggestion
    public fun observeEncryptedLogsUploadResult(): Flow<OnEncryptedLogUploaded?>
```

ParaskP7 replied:
You're right @wzieba , will do! 👍

ParaskP7 replied:
Done: 25195ccc2713d0b506881cee8258a120cab2dd2f @wzieba

ParaskP7 committed:
Api: Expose flow instead of state flow on encrypted logging

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/26
#discussion_r2222788290

ParaskP7 committed:
Refactor: Remove unnecessary fluxc related on changed (error) base class

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/26
#discussion_r2222786377

ParaskP7 committed:
Docs: Update documentation with updated apis

ParaskP7 commented:
FYI: Also a bonus (b2f20581eb45f34d57d61519d6b8961ffbdd514d) @wzieba , was in my TODO list and should have updated that long ago already... 🫣
